base/errorshow.jl

```diff
@@ -1104,7 +1104,27 @@ function fielderror_hint_handler(io, exc)
     end
 end
 
-Experimental.register_error_hint(fielderror_hint_handler, FieldError)
+Experimental.register_error_hint(fielderror_dict_hint_handler, FieldError)
+
+function fielderror_listfields_hint_handler(io, exc)
+    fields = fieldnames(exc.type)
+    if isempty(fields)
+        print(io, "; $(nameof(exc.type)) has no fields at all.")
+    else
+        print(io, ", available fields: $(join(map(k -> "`$k`", fields), ", "))")
+    end
+    props = _propertynames_bytype(exc.type)
+    isnothing(props) && return
+    props = setdiff(props, fields)
+    isempty(props) && return
+    print(io, "\nAvailable properties: $(join(map(k -> "`$k`", props), ", "))")
+end
+
+_extract_val(::Type{Val{V}}) where {V} = V
+_extract_val(::Type{Val}) = nothing
+_propertynames_bytype(::Type{T}) where {T} = promote_op(Val∘propertynames, T) |> _extract_val
```

Comment:
This _extract_val implementation is unsound in several ways, so if we are going to use this implementation, we must fix that:
```suggestion
function _propertynames_bytype(T::Type)
    inferred_names = promote_op(Val∘propertynames, T)
    inferred_names isa DataType && inferred_names <: Val || return nothing
    inferred_names = inferred_names.parameters[1]
    inferred_names isa NTuple{Symbol} || return nothing
    return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
end
```

Reply:
We can also make this much more efficient by adding an early exit to catch most cases
```julia
    which(propertynames, (T,)) === which(propertynames, (Any,)) && return nothing
```

Reply:
@vtjnash, when you object to an implementation, I would prefer if you offer _why_ it is wrong, ideally with examples (e.g. a type where `_propertynames_bytype` throws) in addition to the proposed fix you've given. That way folks understand what is wrong with the original and can learn to be more valuable contributors.

In this case, I'm guessing the problem is that `promote_op` may return any type wider than the true output. For example, `Union{Val, MyUnrelatedType}`.

Would this be a valid fix?

```suggestion
_extract_val(::Type{Val{V}}) where {V} = V
_extract_val(_) = nothing
_propertynames_bytype(::Type{T}) where {T} = promote_op(Val∘propertynames, T) |> _extract_val
```

Reply:
> Not sure what "unsound" means in this context

I think Jameson means that it's unsound in the [logic sense](https://en.wikipedia.org/wiki/Soundness): either it's invalid or it has at least one false premise. In this case I assume Jameson means the latter, i.e. that it makes some assumptions that may not hold in practice.

Since `propertynames` is a generic function, a user may add a method to `propertynames` for objects with a particular type. However, there is no guarantee that what the method returns is consistent with what's expected. As an example, say someone defined `propertynames(::CoolType)` to return a `Vector{Symbol}` rather than a tuple of `Symbol`s. Similarly, `Val(::CoolType) isa Val` may not hold, or the result of `promote_op` may not be a `Val`, as Lilith mentioned. (In fact, even `|>` is a generic function.) Thus, in Jameson's suggestion, he's explicitly checking that the result of the `promote_op` call is in fact a `Val{NTuple{Symbol}}`. These kinds of checks are helpful e.g. to ensure that method definitions don't have unexpected downstream consequences.

> Would this be a valid fix?

My Jameson hat unfortunately does not grant me his omniscience but I think what you suggest would be mostly equivalent to his suggestion with the addition of a type assertion upon unwrapping the `Val`? Like:
```julia
_extract_val(::Type{Val{V}}) where {V} = V::NTuple{Symbol}
```

Reply:
Another benefit of the diff proposed by @vtjnash is less specialization.

Reply:
> with the addition of a type assertion upon unwrapping the `Val`?

I intentionally omitted that because if someone returns an object that quacks like an `NTuple{N, Symbol}` (i.e. supports `isempty`, `map`, and interpolation of its elements) then we should use it.

---

> Another benefit of the diff proposed by @vtjnash is less specialization.

IMO dispatch is a much more elegant way to extract information from types than inspecting type internals.

> Is Base really written robustly to overloads of such fundamental methods? That's honestly impressive.

No, but we do try when it's not too much trouble.

Reply:
> We can also make this much more efficient by adding an early exit to catch most cases

The early exit here _decreases_ performance. It's good to benchmark before recommending this sort of micro-optimization. See https://github.com/JuliaLang/julia/pull/55165#discussion_r1715275750 for benchmarks.

Reply:
The benchmark result just means it doesn't have the other expected fixes compared to the code I suggested above. My version should compile infinitely faster, as it isn't compiled at all

Reply:
I modified the function you recommended by commenting out the line with `NTuple{Symbol}` which should only speed it up.

```julia-repl
julia> using Base: promote_op

julia> function _propertynames_bytype(T::Type)
           inferred_names = promote_op(Val∘propertynames, T)
           inferred_names isa DataType && inferred_names <: Val || return nothing
           inferred_names = inferred_names.parameters[1]
           inferred_names isa NTuple{Symbol} || return nothing
           return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
       end
_propertynames_bytype (generic function with 4 methods)

julia> _propertynames_bytype(Dict)
ERROR: TypeError: in Vararg, in count, expected Int64, got Type{Symbol}
Stacktrace:
 [1] _propertynames_bytype(T::Type)
   @ Main ./REPL[8]:5
 [2] top-level scope
   @ REPL[9]:1
```

When I also add in your recommended optimization it is still _much_ slower than the PR's current implementation (and contains much more direct dependence on internals). I suspect this is be because the function you recommended executes more slowly than compilation.

```
x@x:~$ julia +nightly
A new `nightly` version is available. Install with `juliaup update`.
  o  | Version 1.12.0-DEV.1106 (2024-08-28)
 o o | Commit 644029282cf (2 days old master)
julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> using Base: promote_op

julia> @time @eval _propertynames_bytype(Dict)
  0.011136 seconds (20.98 k allocations: 1019.578 KiB, 66.72% compilation time)
(:slots, :keys, :vals, :ndel, :count, :age, :idxfloor, :maxprobe)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
98.244 ns (0.01 allocs: 0.589 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.002739 seconds (12.35 k allocations: 603.828 KiB, 96.23% compilation time)
(:R, :Q, :p, :P)

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
179.372 ns (0.02 allocs: 0.878 bytes)
x@x:~$ julia +nightly
A new `nightly` version is available. Install with `juliaup update`.
  o  | Version 1.12.0-DEV.1106 (2024-08-28)
 o o | Commit 644029282cf (2 days old master)
julia> function _propertynames_bytype(T::Type)
           which(propertynames, (T,)) === which(propertynames, (Any,)) && return nothing
           inferred_names = promote_op(Val∘propertynames, T)
           inferred_names isa DataType && inferred_names <: Val || return nothing
           inferred_names = inferred_names.parameters[1]
           #inferred_names isa NTuple{Symbol} || return nothing
           return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
       end
_propertynames_bytype (generic function with 1 method)

julia> using Base: promote_op

julia> @time @eval _propertynames_bytype(Dict)
  0.286678 seconds (4.56 M allocations: 251.616 MiB, 21.61% gc time, 98.41% compilation time)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
2.326 μs (23.33 allocs: 985.333 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.032111 seconds (104.80 k allocations: 5.559 MiB, 99.51% compilation time: 18% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
5.844 μs (50 allocs: 2.512 KiB)
```

Reply:
Something seems vaguely suspicious, since that is quite a bit of extra allocation. Especially just noting that my function is type-stable, while yours is testing the performance of only a single signature and re-specializes on every different length of number of fields.
```
julia> @time @eval _propertynames_bytype(Dict)
  0.622177 seconds (4.69 M allocations: 262.580 MiB, 16.18% gc time, 98.57% compilation time)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.088465 seconds (108.28 k allocations: 5.823 MiB, 99.58% compilation time: 22% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.000366 seconds (116 allocations: 5.094 KiB, 0.43% compilation time: 100% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P
```

It appears that the actual issue is that `which` hasn't been compiled yet, so it takes a bit of extra time:
```
julia> which(Int, (Int,));

julia> @time @eval _propertynames_bytype(Dict)
  0.057656 seconds (65.06 k allocations: 3.342 MiB, 83.99% compilation time)
```

So whereas your version takes 11ms to return a new query, my version takes on the order of 0.4ms

Reply:
You benchmarked it taking 0.4ms to rerun the _same_ query, not a new query. In any case, the PR's version is faster on the usecase you posted. Would you please post a reproducible benchmark demonstrating a use case where your proposal is faster than the PR's implementation?

Reply:
It is 0.4ms to run any query that doesn't specialize the `which` method or to rerun any query, as it doesn't run inference for either of those cases. The current PR version is much slower than the former and leaks memory, despite that being the most common case.